doc/reference/trd-dynamic-process-loading.md

```diff
@@ -0,0 +1,326 @@
+required for the new binary before storing it, this method MUST be called before
+calling `write()`. An implementation is responsible for storing individual
+chunks of the process binary.
+
+The `abort()` call deallocates any stored resources and resets the
+implementation to handle a new `setup()` call.
+
+Each operation is asynchronous and must generate a callback if the operation
+returns `Ok(())`.
+
+The interface is intentionally general to support different underlying storage
+formats and storage media.
+
+4.2 Dynamic Process Loading
+---------------------------------
+
+The kernel provides an interface for loading a stored process binary into an
+active Tock process.
+
+```rust
+trait DynamicProcessLoad {
+    /// Request kernel to load the newly stored process.
+    fn load(&self) -> Result<(), ErrorCode>;
+
+    fn set_load_client(&self, client: &'static dyn DynamicProcessLoadClient);
+}
+
+trait DynamicProcessLoadClient {
+    /// The new process has been loaded.
+    fn load_done(&self, result: Result<(), ErrorCode>);
+}
+```
+
+The `load()` operation is asynchronous and must generate a callback if it
+returns `Ok(())`.
+
+This interface does not mandate that the kernel capsule creates a new process.
+Implementations may include a policy for choosing whether to load a new process
+binary. The implementation must also use the board's chosen credential checking
+policy.
+
+
+5 Sequential Process Loading Implementation
+=================================
+
+Tock includes a reference implementation of the dynamic process loading
+architecture. This includes the `AppLoader` capsule and implementations of the
+kernel capsule traits.
+
+The implementations are structs with names that start with `Sequential`. The
+"Sequential" designation indicates that this implementation assumes 1) process
+binaries include TBF headers, 2) process binaries are stored back-to-back, and
+3) process binaries are stored in executable flash that is mapped to the main
+address space of the processor.
+
+This implementation relies on the `SequentialProcessLoaderMachine` which loads
+process binaries from flash, checks their credentials, and verifies uniqueness.
+
+The implementation is structured like this:
+
+```text
+ trait DynamicBinaryStore
+ trait DynamicProcessLoad
+┌────────────────────────────────┐   ┌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┐
+│                                │   ╎                                ╎
+│                              ──┼──►╎                                ╎
+│                                │   ╎                                ╎
+│ SequentialDynamicBinaryStorage │   ╎ SequentialProcessLoaderMachine ╎
+│                                │   ╎                                ╎
+│                                │   ╎                                ╎
+│                                │   ╎                                ╎
+└────────────────────────────────┘   └╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┘
+ hil::NonvolatileStorage
```

Comment:
I don't understand what this diagram is trying to show me? What does it mean that `trait DynamicBinaryStore` is on top of `trait DynamicProcessLoad`?

Reply:
They are both exposed by `SequentialDynamicBinaryStorage`, which itself has a reference to `SequentialProcessLoaderMachine`.